### ISSUE TITLE
ESQL: Performance degradation in dissect benchmark

### ISSUE THREAD

**Issue by craigtaverner** [elastic/elasticsearch#101997]
Between the 25th and 27th of October there was a gap in the benchmarks due to https://github.com/elastic/elasticsearch/pull/101235 causing some queries to fail (https://github.com/elastic/elasticsearch/issues/101455). Once those queries were removed, so benchmarking could continue there was a change in performance for most of the elastic/logs benchmarks for ESQL. Several were improved by some 10% or more, but one in particular regressed a lot.

<img width="835" alt="Screenshot 2023-11-10 at 10 08 26" src="https://github.com/elastic/elasticsearch/assets/114451/8692bc3b-994b-4c2e-a3bb-188865cfc1ee">

The query in question is:

```
FROM logs-postgres*
| DISSECT message "duration: %{query_duration} ms"
| EVAL query_duration_num = TO_DOUBLE(query_duration)
| STATS avg_duration = AVG(query_duration_num)"
```

**Comment by nik9000** [comment]
Jumping back before https://github.com/elastic/elasticsearch/pull/101235 got basically the same performance as now. Digging more.

**Comment by nik9000** [comment]
I jumped back to some code from October 23rd and still got the same performance. I've got some ideas on how to make this faster, but I don't know why it was faster in the past.

**Comment by craigtaverner** [comment]
Just to check, I've reviewed all changes to rally-tracks a few days before to a few days after this performance degradation, in case there was a change to the actual query that could cause this, but see nothing like that. There were many changes around this time to queries adjacent to this one (mostly adding new queries and temporarily removing and re-adding others), but nothing appears to have changed with this particular query around that time.

The query right now:

```
    {
      "name": "esql_dissect_duration_and_stats",
      "description": "Based on observability queries for duration average",
      "operation-type": "esql",
      "query": "FROM logs-postgres* | DISSECT message \"duration: %{query_duration} ms\" | EVAL query_duration_num = TO_DOUBLE(query_duration) | STATS avg_duration = AVG(query_duration_num)"
    },
```

Git blame says the ESQL query itself is unchanged since it was added by me on Oct 9th. The biggest change was the introduction of the `"operation-type": "esql"` line on October 19th, by Brad, which he did to most ESQL queries at that time. This did not affect the performance as it ran like that for a few days before the regression.

Is it possible that a change in `rally` itself could account for this? Why just this query though?

**Comment by craigtaverner** [comment]
This seems to have been fixed on the 30th November, and now more recently we see another improvement. I think we can close this issue.
<img width="793" alt="Screenshot 2024-01-08 at 12 32 56" src="https://github.com/elastic/elasticsearch/assets/114451/54ee1699-9f93-4e70-a70a-785a58800577">

**Comment by nik9000** [comment]
I think the recent fix is #103632. I think the 30th of November one was something @dnhatn did.
